File: src/add_two_numbers/add_two_numbers.py
#!/usr/bin/env python
# coding:utf-8
# filename:add_two_numbers.py

'''
author:	gavingeng
date:	2018-05-31 01:01:22 
'''


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        '''
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        '''
        retNode = ListNode(-1)
        curNode = retNode

        carry = 0

        while (l1 != None or l2 != None):
            x = 0 if l1 == None else l1.val
            y = 0 if l2 == None else l2.val
            sum = carry + x + y
            curNode.next = ListNode(sum % 10)
            carry = sum / 10
            curNode = curNode.next
            l1 = None if l1 == None else l1.next
            l2 = None if l2 == None else l2.next

        return retNode.next

    def show(self, node):
        while (node != None):
            print node.val
            node = node.next


def test():
    l1 = ListNode(2)
    l1.next = ListNode(4)
    l1.next.next = ListNode(3)

    l2 = ListNode(5)
    l2.next = ListNode(6)
    l2.next.next = ListNode(4)
    l2.next.next.next = ListNode(9)

    s = Solution()
    node = s.addTwoNumbers(l1, l2)
    s.show(node)


def main():
    test()


if __name__ == '__main__':
    main()

File: src/add_two_numbers/addTwoNumbers.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	retNode := ListNode{Val: -1}
	curNode := retNode

	carry := 0

	for l1 != nil || l2 != nil {
		x := 0
		if l1 != nil {
			x = l1.Val
		}

		y := 0
		if l2 != nil {
			y = l2.Val
		}
		sum := carry + x + y
		curNode.Next = &ListNode{Val: sum % 10}
		carry = sum / 10
		curNode = *curNode.Next
		if l1 != nil {
			l1 = l1.Next
		} else {
			l1 = nil
		}
		if l2 != nil {
			l2 = l2.Next
		} else {
			l2 = nil
		}
	}

	if carry > 0 {
		curNode.Next = &ListNode{Val: carry}
	}

	return retNode.Next
}

func show(node *ListNode) {
	for node != nil {
		fmt.Println(node.Val)
		node = node.Next
	}
}

func test() {
	l1 := ListNode{Val: 2, Next: nil}
	l1.Next = &ListNode{Val: 4, Next: nil}
	l1.Next.Next = &ListNode{Val: 3, Next: nil}

	l2 := ListNode{Val: 5, Next: nil}
	l2.Next = &ListNode{Val: 6, Next: nil}
	l2.Next.Next = &ListNode{Val: 4, Next: nil}
	l2.Next.Next.Next = &ListNode{Val: 9, Next: nil}

	node := addTwoNumbers(&l1, &l2)
	show(node)
}

func test1() {
	l1 := ListNode{Val: 5, Next: nil}
	l2 := ListNode{Val: 5, Next: nil}

	node := addTwoNumbers(&l1, &l2)
	show(node)
}

func main() {
	test()
	//test1()
}
